gossip/hub/handlers.go

```diff
@@ -0,0 +1,664 @@
+	// Send request to the Log server.
+	req := trillian.GetLatestSignedLogRootRequest{LogId: logID}
+	glog.V(2).Infof("%s: GetLogRoot => grpc.GetLatestSignedLogRoot %+v", prefix, req)
+	rsp, err := client.GetLatestSignedLogRoot(ctx, &req)
+	glog.V(2).Infof("%s: GetLogRoot <= grpc.GetLatestSignedLogRoot err=%v", prefix, err)
+	if err != nil {
+		return nil, fmt.Errorf("backend GetLatestSignedLogRoot request failed: %v", err)
+	}
+
+	// Check over the response.
+	slr := rsp.SignedLogRoot
+	if slr == nil {
+		return nil, errors.New("no log root returned")
+	}
+	// TODO(drysdale): configure the hub with the Trillian public key for this tree, and check the Trillian signature here.
+	var root types.LogRootV1
+	if err := root.UnmarshalBinary(slr.LogRoot); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal log root: %v", err)
+	}
+
+	glog.V(3).Infof("%s: GetLogRoot <= root=%+v", prefix, root)
+	return &root, nil
+}
+
+func getSTH(ctx context.Context, c *hubInfo, w http.ResponseWriter, r *http.Request) (int, error) {
+	root, err := GetLogRoot(ctx, c.rpcClient, c.logID, c.hubPrefix)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	hth := api.HubTreeHead{
+		TreeSize:  root.TreeSize,
+		Timestamp: root.TimestampNanos, // in nanoseconds since UNIX epoch
+		RootHash:  root.RootHash,
+	}
+	hthData, err := tls.Marshal(hth)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to marshal tree head: %v", err)
+	}
+	signature, err := c.signer.Sign(hthData)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to sign STH data: %v", err)
+	}
+
+	// Now build the final result object and marshal to JSON
+	jsonRsp := api.GetSTHResponse{TreeHeadData: hthData, HubSignature: signature}
+	jsonData, err := json.Marshal(&jsonRsp)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to marshal response: %v %v", jsonRsp, err)
+	}
+
+	w.Header().Set(contentTypeHeader, contentTypeJSON)
+	if _, err := w.Write(jsonData); err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to write response data: %v", err)
+	}
+
+	return http.StatusOK, nil
+}
+
+func getSTHConsistency(ctx context.Context, c *hubInfo, w http.ResponseWriter, r *http.Request) (int, error) {
+	firstVal := r.FormValue(api.GetSTHConsistencyFirst)
+	if firstVal == "" {
+		return http.StatusBadRequest, errors.New("parameter 'first' is required")
+	}
+	secondVal := r.FormValue(api.GetSTHConsistencySecond)
+	if secondVal == "" {
+		return http.StatusBadRequest, errors.New("parameter 'second' is required")
+	}
+
+	first, err := strconv.ParseInt(firstVal, 10, 64)
+	if err != nil {
+		return http.StatusBadRequest, errors.New("parameter 'first' is malformed")
+	}
+	second, err := strconv.ParseInt(secondVal, 10, 64)
+	if err != nil {
+		return http.StatusBadRequest, errors.New("parameter 'second' is malformed")
+	}
+
+	if first < 0 || second < 0 {
```

Comment:
How much parsing / validation is needed here? Is it a duplication of what's in Trillian?

Reply:
Yep, it's mostly a duplicate of `validateGetConsistencyProofRequest`, but checking here saves an unnecessary RPC.